src/util/numbering.h

```diff
@@ -16,154 +16,82 @@ Author: Daniel Kroening, kroening@kroening.com
-      INVARIANT(data.size()==numbers.size(), "vector sizes must match");
+      data_.emplace_back(a);
+      INVARIANT(data_.size() == numbers_.size(), "vector sizes must match");
     }
 
     return (result.first)->second;
   }
 
-  number_type operator()(const T &a)
-  {
-    return number(a);
-  }
-
-  bool get_number(const T &a, number_type &n) const
-  {
-    typename numberst::const_iterator it=numbers.find(a);
-
-    if(it==numbers.end())
-      return true;
-
-    n=it->second;
-    return false;
-  }
-
-  void clear()
+  optionalt<number_type> get_number(const key_type &a) const
   {
-    data.clear();
-    numbers.clear();
-  }
-
-  size_t size() const { return data.size(); }
-
-  T &operator[](size_type t) { return data[t]; }
-  const T &operator[](size_type t) const { return data[t]; }
-
-  iterator begin() { return data.begin(); }
-  const_iterator begin() const { return data.begin(); }
-  const_iterator cbegin() const { return data.cbegin(); }
-
-  iterator end() { return data.end(); }
-  const_iterator end() const { return data.end(); }
-  const_iterator cend() const { return data.cend(); }
-};
-
-template <typename T, class hash_fkt>
-// NOLINTNEXTLINE(readability/identifiers)
-class hash_numbering final
-{
-public:
-  // NOLINTNEXTLINE(readability/identifiers)
-  typedef unsigned int number_type;
-
-private:
-  typedef std::vector<T> data_typet;
-  data_typet data;
-  typedef std::unordered_map<T, number_type, hash_fkt> numberst;
-  numberst numbers;
-
-public:
-  // NOLINTNEXTLINE(readability/identifiers)
-  typedef typename data_typet::size_type size_type;
-  // NOLINTNEXTLINE(readability/identifiers)
-  typedef typename data_typet::iterator iterator;
-  // NOLINTNEXTLINE(readability/identifiers)
-  typedef typename data_typet::const_iterator const_iterator;
-
-  number_type number(const T &a)
-  {
-    std::pair<typename numberst::const_iterator, bool> result=
-      numbers.insert(
-      std::pair<T, number_type>
-      (a, number_type(numbers.size())));
-
-    if(result.second) // inserted?
+    const auto it = numbers_.find(a);
+    if(it == numbers_.end())
     {
-      this->push_back(a);
-      assert(this->size()==numbers.size());
+      return {};
     }
-
-    return (result.first)->second;
-  }
-
-  bool get_number(const T &a, number_type &n) const
-  {
-    typename numberst::const_iterator it=numbers.find(a);
-
-    if(it==numbers.end())
-      return true;
-
-    n=it->second;
-    return false;
+    return it->second;
   }
 
   void clear()
   {
-    data.clear();
-    numbers.clear();
+    data_.clear();
+    numbers_.clear();
   }
 
-  template <typename U>
-  void push_back(U &&u) { data.push_back(std::forward<U>(u)); }
+  // clang-format off
+  size_t size() const { return data_.size(); }
 
-  T &operator[](size_type t) { return data[t]; }
-  const T &operator[](size_type t) const { return data[t]; }
+  template <typename U>
+  void push_back(U &&u) { data_.push_back(std::forward<U>(u)); }
```

Comment:
Does this need the explicit template param?

Reply:
Yes. There's tw overloads of `push_back` in `vector`. This allows us to use both of them from a single wrapper function, rather than providing separate overloads for rvalue refs and const lvalue refs.

Reply:
Yeah, I mean surely `std::forward(u)` would do?

Reply:
Oh wait I see, ok ignore this